text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,482 @@
+
+A "Receive" span covers such calls and should link to the "Create" spans of all
+messages that are forwarded via the respective call. Depending on the use case,
+"Receive" spans can correlate with "Process" spans or other spans modelling
+processing operations.
+
+#### General considerations for both push-based and pull-based scenarios
+
+The operations modelled by "Deliver" or "Receive" spans do not strictly refer
+to receiving the message from intermediaries, but instead refer to the
+application receiving messages for processing. If messages are fetched from the
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Deliver" or "Receive" span. However, libraries or SDKs
+might pre-fetch messages from intermediaries and cache those messages, and only
+forward messages to the application at a later time. In this case, the
+operation of pre-fetching and caching should not be covered by the "Deliver" or
+"Receive" spans.
+
+Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
+to notify the application that no message is available for processing), one
+message, or multiple messages (a batch of messages). "Deliver" and "Receive"
+spans should link to the "Create" span of the messages forwarded, thus those
+spans can link to zero, one, or multiple "Create" spans.
+
+#### Settlement
+
+Messages can be settled in a variety of different ways. In some cases, messages
+are not settled at all (fire-and-forget), in other cases settlement operations
+are triggered manually by the user, and in callback scenarios settlement can be
+automatically triggered by messaging SDKs based on return values of callbacks.
+
+A "Settle" span should be created for every settlement operation, no matter if
+the settlement operation was manually triggered by the user or automatically
+triggered by SDKs. SDKs will in some cases auto-settle messages in
+push-scenarios when messages are delivered via callbacks. In this case, it is
+recommended to create a parent span, so that a "Settle" span will be a sibling
+of the related "Deliver" span.
+
+Alternatively, an event can be created instead of a "Settle" span. Events could
+be added to "Deliver" spans or to ambient spans.
+
+"Settle" spans may link to "Create" spans of the messages that are settled,
+however, for some settlement scenarios this is not feasible or possible.
+
+## Proposed changes and additions to the messaging semantic conventions
+
+### Operation name
+
+The following operations related to messages are covered by these semantic
+conventions:
+
+| Operation name | Description |
+|----------------|-------------|
+| `publish`      | One or more messages are provided for publishing to an intermediary. |
+| `create`       | A message is created. |
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more messages are passed to a consumer. |
+| `settle`       | One or more messages are settled. |
```

Comment:
Should we suggest composite names like?

create_publish
receive_process
...

Reply:
What benefit would one get from it? It's not clear to me.

Reply:
because there are many cases where a message is created and published inline... Unless we point to Create being used by app code and publish to the lib actually managing the connection. 
If this is the case, it would be a nice clarification.

Reply:
For the "create_publish" combination, this is already covered by the supported case of a "publish" span that is of kind "PRODUCER". In this case we know that no "create" spans are present.

My stance is, that we shouldn't suggest composite names for other cases. Firstly, because a set of five distinct operations is much easier to grasp than a set of operations enlarged by composites. If there's "receive_process", there's probably a "deliver_process", and the people might ask for "receive_process_settle". Secondly, if you collapse operations, like e. g. "receive_process", you lose the possibility to model cases where the "receive" operation succeeds, but the "process" operation fails.

However, I'm open towards other opinions.

Reply:
This was discussed in today's messaging workgroup. There are cases where composite names would make sense, however, we're not sure if there are enough use-cases to merit the additional complexity.

We decided to stick with simple operation names for now, as composite names can be added on top of that in the future if there's need for it (whereas it's much harder the harder way around).